src/RoslynInsertionTool/RoslynInsertionTool/RoslynInsertionTool.VisualStudioTeamServices.cs

```diff
@@ -222,6 +236,43 @@ private static GitPullRequest CreatePullRequest(string sourceBranch, string targ
             throw new InvalidOperationException("Could not find drop path");
         }
 
+        private static async Task<string> DownloadArtifactsAsync(BuildHttpClient buildClient, Build build, BuildArtifact artifact, CancellationToken cancellationToken)
+        {
+            var tempDirectory = Path.Combine(Path.GetTempPath(), string.Concat(Options.InsertionName, Options.BranchName).Replace(" ", "_").Replace("/","_"));
+            if (Directory.Exists(tempDirectory))
+            {
+                // Be judicious and clean up old artifacts so we do not eat up memory on the scheduler machine.
+                Directory.Delete(tempDirectory, recursive: true);
+                // Sometimes a creation of a directory races with deletion since at least in .net 4.6 deletion is not a blocking call.
+                // Hence explictly waiting for the directory to be deleted before moving on.
+                Stopwatch w = Stopwatch.StartNew();
+                while (Directory.Exists(tempDirectory) && w.ElapsedMilliseconds < 10 * 1000) Thread.Sleep(100);
```

Comment:
nit: Could you break this up into multiply lines

It might also be good to have this use a back-off factor and a retry count, rather than just sleeping for 100ms intervals.

Having 10 retries, a starting wait time of 10ms, and a back-off factor of 2 will give us ~10s to complete 10 tries.

Reply:
So the Directory.Delete itself does not have any problems. Its not that it does not delete the directory, its just that it does not necessarily delete it immediately after it comes out of that call. The check for existence should ensure that condition holds true, hence dont see a need for a retry as such. I could add a back-off factor but if there isn't a retry it wouldn't make much of a difference.

Reply:
@tannergooding : Fixed the spacing.